### PR TITLE
Migrate Hashi SDK into the monorepo as @mysten/hashi

### PR DESCRIPTION
**Pull request by Tzal3x**
## Description

Migrates the Hashi TypeScript SDK (native BTC collateralization on Sui) from the standalone
[hashi-ts-sdk](https://github.com/MystenLabs/hashi-ts-sdk) repo into `packages/hashi`, renaming it
from `@mysten-incubation/hashi` (v0.5.0 on npm) to `@mysten/hashi` (SEDEFI-163).

- **Source, tests, and committed codegen bindings** copied from the standalone repo and reformatted
  to monorepo prettier style (tabs, single quotes, import sort) — most of the diff is this
  mechanical reformat plus the generated `src/contracts/**` bindings.
- **package.json** rewritten to monorepo conventions: `@mysten/sui` as a `workspace:^` peer
  dependency, `@mysten/codegen` as a workspace devDependency, standard build/lint/test scripts.
- **Codegen** (`sui-codegen.config.ts`) retargeted at the `../hashi` sibling checkout
  (walrus/payment-kit pattern); bindings stay committed at the revision generated from
  MystenLabs/hashi `cd2b81f`.
- **Monorepo wiring**: `release-hashi.yml` workflow (OIDC trusted publishing via
  `_release-package.yml`), oxlint ignore for `packages/hashi/src/contracts`,
  `linguist-generated` for `src/contracts/**`, `AGENTS.md` sibling-repo/codegen entries, and
  minor changesets — so the first publish under the new name is 0.6.0 with changelog entries
  explaining the rename and the breaking API change below.
- **CI**: unit tests run via the standard `test` task; `hashi-ci.yml` ports the standalone repo's
  localnet integration harness — it clones `MystenLabs/hashi` (pinned to `cd2b81f`), builds the
  `hashi-localnet` Rust CLI, installs a version-pinned, SHA256-verified Bitcoin Core, and runs the
  env-gated integration tests (`HASHI_E2E_*`) against a live localnet.
- **Test type fixes**: the new `test:typecheck` step surfaced latent type errors the standalone
  repo never caught (it never ran `tsc` over tests): unnarrowed `TransactionHistoryItem` union
  access in three integration tests, and a parse-only `$kind` tag in a BCS enum serialize input.
- Docs are README-only for now (the standalone repo's comprehensive guide travels as the package
  README); a docs-site section can follow at graduation.

### Review follow-ups (post-migration commits)

API changes from review:

- **Breaking**: `hashi()` no longer takes a `network` option — the network is derived from the Sui
  client being extended (`.$extend(hashi())`), still throwing for networks without a
  `NETWORK_CONFIG` entry unless custom `hashiObjectId`/`packageId` are passed.
- `HashiClientOptions<Name>` defaults to `string` instead of a wrong `'HashiClient'` literal.

Internal cleanups from review (no API change):

- Withdrawal building uses `tx.balance(...)` instead of hand-composing `coinWithBalance` +
  `0x2::coin::into_balance`.
- Gas estimators reuse the real `tx.deposit`/`tx.requestWithdrawal` builders instead of
  duplicating the PTBs, so estimates can't drift from what users sign.
- Dropped the manual 500-object chunking (`core.getObjects` batches internally on all three
  transports) and replaced the raw string-interpolated GraphQL fetch with a `SuiGraphQLClient`
  query using proper GraphQL variables.
- `rpcCall<T>` is generic instead of casting at each Bitcoin RPC call site.

## Test plan

- `pnpm turbo build --filter @mysten/hashi` — typecheck + tsdown pass against workspace
  `@mysten/sui` 2.20.x (verifies the peer-range move from `^2.14.1`)
- `pnpm --filter @mysten/hashi test` — test-dir typecheck + 185 unit tests pass
- `pnpm turbo lint --filter @mysten/hashi` + root oxlint — clean (0 findings on ported code)
- `pnpm changeset status` — `@mysten/hashi` queued for a minor bump (0.6.0)
- Codegen exercised end-to-end against the `../hashi` sibling checkout (regenerates successfully;
  regenerated output not committed since that checkout is ahead of the pinned contract revision)
- `Localnet Integration` CI job green — full deposit/withdrawal integration flow against a live
  localnet with pinned Bitcoin Core

Post-merge follow-ups (outside this PR): dispatch `release-hashi.yml` (dry-run first), deprecate
`@mysten-incubation/hashi` on npm, archive the hashi-ts-sdk repo.

---

### AI Assistance Notice

> Please disclose the usage of AI. This is primarily to help inform reviewers of how careful they need to review PRs, and to keep track of AI usage across our team. Please fill this out accurately, and do not modify the content or heading for this section!

- [x] This PR was primarily written by AI.
- [ ] I used AI for docs / tests, but manually wrote the source code.
- [ ] I used AI to understand the problem space / repository.
- [ ] I did not use AI for this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Tzal3x** [comment]
The Turborepo CI failure is unrelated to this PR: the `Lint, Build, and Test` job passes, and only the `pnpm audit` job fails because npm has retired the legacy audit endpoints (they now return HTTP 410), which the pinned pnpm version still calls — a repo-wide infra issue affecting all branches, not this change.

File: packages/hashi/src/contracts/hashi/deps/sui/group_ops.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/**************************************************************
```
Comment (hayes-mysten) [inline packages/hashi/src/contracts/hashi/deps/sui/group_ops.ts:1]: these generated files should be marked as generated via linguist configs so they auto-collapse in reviews

File: packages/hashi/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+export type SuiNetwork = 'devnet' | 'testnet' | 'mainnet' | 'localnet';
+
+export interface NetworkConfig {
+	hashiObjectId: string;
+	packageId: string;
+	bitcoinNetwork: BitcoinNetwork;
+}
+
+export interface HashiClientOptions<Name = 'HashiClient'> {
```
Comment (hayes-mysten) [inline packages/hashi/src/types.ts]: the default Name here seems wrong, I assume this is meant to be `hashi`

I think actually `string` is probably a better default here.  If something accepts `HashiClientOptions` only `string` would allow any options object with different names set

File: packages/hashi/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+	name?: Name;
+	/** Sui network — determines Hashi object IDs and default Bitcoin network. */
+	network: SuiNetwork;
```
Comment (hayes-mysten) [inline packages/hashi/src/types.ts]: we generally don't want to pass network separately, and instead get it from the client being extended

File: packages/hashi/src/btc-rpc.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { UtxoLookupResult } from './types.js';
+
+async function rpcCall(
```
Comment (hayes-mysten) [inline packages/hashi/src/btc-rpc.ts]: I like just making these kinds of methods take a generic so you can pass in the data type rather than always casting after calling it

File: packages/hashi/src/client.ts [hunk 5]
```diff
@@ -0,0 +1,1635 @@
...
+		hashiObjectId,
+		packageId,
+		bitcoinNetwork,
+		btcRpcUrl,
+		graphqlUrl,
+		guardianUrl,
+		guardianInfoProvider,
+	}: {
+		client: ClientWithCoreApi;
+		network: SuiNetwork;
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: network should always come from the client, and we should throw if its unsupported

File: packages/hashi/src/client.ts [hunk 6]
```diff
@@ -0,0 +1,1635 @@
...
+			throw new Error(
+				`Hashi is not yet supported on Sui ${network}. Provide a custom hashiObjectId and packageId.`,
+			);
+		}
+		this.#client = client;
+		this.#hashiObjectId = resolvedObjectId;
+		this.#packageId = resolvedPackageId;
+		this.#bitcoinNetwork = bitcoinNetwork ?? config?.bitcoinNetwork ?? 'testnet';
+		this.#btcRpcUrl = btcRpcUrl;
+		this.#graphqlUrl = graphqlUrl ?? defaultGraphqlUrl(network);
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: why not just create a graphql client here

File: packages/hashi/src/client.ts [hunk 7]
```diff
@@ -0,0 +1,1635 @@
...
+			/**
+			 * Target Bitcoin address as raw witness program bytes — 20 bytes for
+			 * P2WPKH, 32 bytes for P2TR. Callers decode their own bech32(m)
+			 * strings for now; a string-input overload may land in a follow-up.
+			 */
+			bitcoinAddress: Uint8Array;
+		}): Transaction => {
+			const tx = new Transaction();
+			const btcType = `${this.#packageId}::btc::BTC`;
+			const coin = tx.add(
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: we've added tx.coin and tx.balance (tx.coin is just an alias for coinWithBalance).

You also don't need to tx.add explicitly here you can just do `const coin = conWithBalance() || tx.coin(..)`.

File: packages/hashi/src/client.ts [hunk 8]
```diff
@@ -0,0 +1,1635 @@
...
+				coinWithBalance({
+					type: btcType,
+					balance: options.amount,
+					useGasCoin: false,
+				}),
+			);
+			const [balance] = tx.moveCall({
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: if you are using a balance directly you should just use `tx.balance(...)`

File: packages/hashi/src/client.ts [hunk 9]
```diff
@@ -0,0 +1,1635 @@
...
+				package: '0x2',
+				module: 'coin',
+				function: 'into_balance',
+				typeArguments: [btcType],
+				arguments: [coin],
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: intents like coins and balances can also be used dirctly in arguments (`arguments: [tx.coin()])` (splitting them out is totally fine, but if they are only being used in one place it can make it easier to read in some cases (this is very subjective and I don't have strong opinions on how to do it)

File: packages/hashi/src/client.ts [hunk 10]
```diff
@@ -0,0 +1,1635 @@
...
+			const tx = new Transaction();
+			const balance = tx.add(this.call.cancelWithdrawal({ requestId: options.requestId }));
+			const [coin] = tx.moveCall({
+				package: '0x2',
+				module: 'coin',
+				function: 'from_balance',
+				typeArguments: [`${this.#packageId}::btc::BTC`],
+				arguments: [balance],
+			});
+			tx.transferObjects([coin], options.recipient);
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts:587]: I am curious about plans for if/when we start just sending users balances instead of coins

File: packages/hashi/src/client.ts [hunk 11]
```diff
@@ -0,0 +1,1635 @@
...
+		 * @param sender - If provided, estimates gas cost via dry-run.
+		 */
+		withdrawalFees: async (sender?: string): Promise<WithdrawalFees> => {
+			const snap = await this.view.all();
+
+			let gasEstimateMist = 0n;
+			if (sender) {
+				const dummyAmount = snap.bitcoinWithdrawalMinimum + 1n;
+				const btcType = `${this.#packageId}::btc::BTC`;
+				const tx = new Transaction();
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: Are these methods just building exactly the same transactions as the `client.tx` transactions?  This seems like we should be reusing those to get accurate gas estimates that stay in sync without manually keeping them in sync with changes

File: packages/hashi/src/client.ts [hunk 12]
```diff
@@ -0,0 +1,1635 @@
...
+	 * Resolve the Bag id holding a user's request IDs from `user_requests`,
+	 * or `null` if the user has never had a request.
+	 */
+	async #fetchUserRequestsBagId(tableId: string, suiAddress: string): Promise<string | null> {
+		try {
+			const { dynamicField } = await this.#client.core.getDynamicField({
+				parentId: tableId,
+				name: {
+					type: 'address',
+					bcs: bcs.Address.serialize(suiAddress).toBytes(),
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts:1399]: we added a top level toBytes method to avoid the extra chained serialize call

File: packages/hashi/src/client.ts [hunk 13]
```diff
@@ -0,0 +1,1635 @@
...
+		} while (cursor);
+		return keys;
+	}
+
+	/**
+	 * Batch-fetches objects in chunks of `GET_OBJECTS_BATCH`, concatenating
+	 * the results. Each element is either an object or an `Error` (for
+	 * missing / deleted objects).
+	 */
+	async #batchGetObjects(objectIds: string[], include?: { content?: boolean }) {
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: the core getObjects API already handles batching internally

File: packages/hashi/src/client.ts [hunk 14]
```diff
@@ -0,0 +1,1635 @@
...
+                events(
+                    filter: { sender: "${sender}", type: "${eventType}" }
+                    first: 50${afterClause}
+                ) {
+                    nodes { contents { json } }
+                    pageInfo { hasNextPage endCursor }
+                }
+            }`;
+
+			const res = await fetch(this.#graphqlUrl, {
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts]: this should use the graphql client, which will properly type all the graphql requests for you

Comment (hayes-mysten) [reply packages/hashi/src/client.ts]: we are also a couple weeks away from having event queries work in grpc and the core APIs

**Review by hayes-mysten — COMMENTED** [review]
Overall this looks really good and aligns well with this repos best practices.   Code looks clean and well organized.

I've gone over everything from a best practices and sdk usage perspective and it all looks pretty solid to me.  I did not review it for correctness, edge cases, etc, and I did not really look at the tests or docs.

**Comment by hayes-mysten** [comment]
@clud-bot review

File: packages/hashi/sui-codegen.config.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import type { SuiCodegenConfig } from '@mysten/codegen';
+
+const config: SuiCodegenConfig = {
+	output: './src/contracts',
+	packages: [
+		{
+			package: '@local-pkg/hashi', // TODO: update this when hashi is published on MVR.
+			path: '../../../hashi/packages/hashi',
```
[the remark this answers is not in the document]
Comment (Tzal3x) [reply packages/hashi/sui-codegen.config.ts:11]: Skipping this one — I think the prerequisite is already documented in the place you pointed at.

`AGENTS.md` gained a `../hashi` row in its "Local Sibling Repositories" table as part of this PR:

| `../hashi` | Hashi BTC-bridge Move contracts | `@mysten/hashi` codegen |

And none of the five sibling configs (`walrus`, `payment-kit`, `deepbook-v3`, `pas`, `suins-v2`) carry an inline comment either — they all use a bare `../../../<repo>/...` path. So the established pattern is "path in the config, prose in AGENTS.md", which hashi already follows; adding a comment only here would make it the asymmetric one.

Happy to revisit as a convention, but it should land as its own change across all six configs rather than just this one.

File: packages/hashi/src/contracts/hashi/deps/sui/group_ops.ts (hunk 1 above)
Comment (Tzal3x) [reply packages/hashi/src/contracts/hashi/deps/sui/group_ops.ts:1]: Fixed in 0876689a — added `packages/hashi/src/contracts/** linguist-generated=true` to the root `.gitattributes`, alongside the existing entries for walrus, payment-kit, deepbook-v3, pas, etc. Note this only affects diff rendering on future PRs; the files will still show expanded in this one.

File: packages/hashi/CHANGELOG.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+### Minor Changes
+
+- 75fcdca: Fix btcTxid display values to strip the 0x prefix. Add GraphQL-based discovery of pending deposits to transaction history — confirmed requests still read from the on-chain user_requests index; in-flight deposits are discovered via DepositRequestedEvent queries and deduplicated. Bump GET_OBJECTS_BATCH to 500.
+
+## 0.0.2
+
+### Patch Changes
+
+- 9422708: Add a package-level `README.md` so the npm landing page has a real overview (install, one quickstart snippet, link to the repo README for full docs). Also corrects stale `@mysten/hashi` references in the root README to the actual published name `@mysten/hashi`.
```
[the remark this answers is not in the document]
Comment (Tzal3x) [reply packages/hashi/CHANGELOG.md]: Fixed in b68aede9. Checked the standalone repo history — the original entry read "... to the actual published name `@mysten-incubation/hashi`", and the migration ran a blanket `@mysten-incubation/hashi` → `@mysten/hashi` rename that collapsed the second occurrence. So the fix belongs on the trailing name, not the leading one; correcting the first would have inverted the historical fact. Diffed the full CHANGELOG against the original and this was the only line the rename damaged.

File: packages/hashi/test/integration/_env.ts [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+		throw new Error(
+			'No signer available: set HASHI_E2E_FUNDED_KEY_PEM (localnet, written ' +
+				'by `hashi-localnet start`) or HASHI_E2E_SUI_PRIVATE_KEY (devnet, bech32).',
+		);
+	}
+	return Ed25519Keypair.fromSecretKey(decodeSuiPrivateKey(bech32).secretKey);
+}
+
+/**
+ * Invokes the `hashi-localnet` Rust CLI from the submodule. The CI workflow
```
[the remark this answers is not in the document]
Comment (Tzal3x) [reply packages/hashi/test/integration/_env.ts]: Fixed in 5922f2b9 — "from the submodule" → "from the hashi repo checkout", matching `hashi-ci.yml` which clones `MystenLabs/hashi` into `hashi/` and builds the binary there. Also grepped for other stale references; the remaining "submodule" hits in the workflow are correct (two explain that ts-sdks has no submodule, one is `submodules: recursive` on the hashi checkout itself).

File: .github/workflows/hashi-ci.yml [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+            sed -n 's/.*"tag_name": "v\([^"]*\)".*/\1/p' || echo "$FALLBACK_VERSION")
+
+          if [ -z "$BITCOIN_VERSION" ]; then
+            echo "Failed to detect version, using fallback: $FALLBACK_VERSION"
+            BITCOIN_VERSION="$FALLBACK_VERSION"
+          fi
+
+          echo "Installing Bitcoin Core v${BITCOIN_VERSION}..."
+
+          wget -q https://bitcoincore.org/bin/bitcoin-core-${BITCOIN_VERSION}/bitcoin-${BITCOIN_VERSION}-x86_64-linux-gnu.tar.gz || {
```
[the remark this answers is not in the document]
Comment (Tzal3x) [reply .github/workflows/hashi-ci.yml]: Fixed in 460a075d, and went a bit further than the suggestion.

Reading the whole step, the tarball was not just unverified — the version was auto-detected from the GitHub releases API, so CI silently tracked whatever upstream tagged most recently. That is its own reproducibility problem, and it also blocks checksum pinning, since you cannot pin a hash for a version you do not know in advance.

Also worth noting: fetching SHA256SUMS from bitcoincore.org to verify a tarball from bitcoincore.org is a same-origin check — it catches truncation and network MITM, but not a compromise of the host serving both.

So the step now pins `BITCOIN_VERSION: 29.1` plus its SHA256 and runs `sha256sum -c -` before `tar`, and the auto-detect and fallback-retry are gone (net -8 lines). This also brings the step in line with the rest of the workflow, which SHA-pins actions and pins the hashi clone.

Verified both directions locally: the real tarball passes, a tampered hash exits 1 before extraction. Tradeoff is that picking up new bitcoind releases is now a manual two-line bump; the step comment says where to get the checksum.

File: packages/hashi/CHANGELOG.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+## 0.5.0
+
+### Minor Changes
+
+- ff9a398: Wire up Sui testnet: add the testnet Hashi object and package ids to `NETWORK_CONFIG` (BTC signet), so `hashi({ network: "testnet" })` works out of the box
+
+### Patch Changes
+
+- ff9a398: Bump hashi submodule to cd2b81f (no contract or binding changes)
```
[the remark this answers is not in the document]
Comment (Tzal3x) [reply packages/hashi/CHANGELOG.md:11]: Leaving this one as-is — `hashi-ts-sdk/.gitmodules` confirms the standalone repo did have a `hashi` submodule, so this historical release note was accurate when written and the monorepo layout does not apply retroactively to it. (Contrast with the `_env.ts` comment, which describes current behaviour and I did fix.) Happy to change it if you would rather changelog entries read against the current repo layout.

File: packages/hashi/src/types.ts (hunk 2 above)
Comment (Tzal3x) [reply packages/hashi/src/types.ts]: Fixed in d9bd8688 — changed the default to `string`.

File: packages/hashi/src/btc-rpc.ts (hunk 4 above)
Comment (Tzal3x) [reply packages/hashi/src/btc-rpc.ts]: Fixed in 38b3d35a — rpcCall<T> now takes the type param instead of casting at each call site.

File: packages/hashi/src/client.ts (hunk 12 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts:1399]: I couldn't find a top-level toBytes in @mysten/bcs or @mysten/sui/bcs — BcsType has no toBytes(value), only the no-arg SerializedBcs.toBytes() after .serialize(). Checked origin/main too, no newer commit adds one. Can you point me at where it lives? Leaving this open until then.

File: packages/hashi/src/client.ts (hunk 10 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts:587]: The teams that will be using this sdk will be defi builders. I have the impression that it would be more useful providing them with that kind of granularity. 

On the other hand sending the user balance directly seems very convenient. Maybe we could add this as extra functionality instead of _replacing_ the coins.

File: packages/hashi/src/types.ts (hunk 3 above)
Comment (Tzal3x) [reply packages/hashi/src/types.ts]: Fixed in 3bf25ce8 — `network` is removed from `HashiClientOptions`; the constructor now reads `client.network` from the extended client. Added a major changeset since this changes the public API.

File: packages/hashi/src/client.ts (hunk 5 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 3bf25ce8 — network now always comes from `client.network`, and the existing unsupported-network throw is kept (still bypassable with custom `hashiObjectId`/`packageId` for local deployments).

File: packages/hashi/src/client.ts (hunk 7 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 82cd7467 — went one step further: since the coin was only created to feed `into_balance`, the whole pair collapses into a single `tx.balance(...)` (the intent resolver inserts the `into_balance` conversion at build time). Also applied at the gas-estimate site.

File: packages/hashi/src/client.ts (hunk 8 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 82cd7467 — `tx.balance(...)` replaces the coin + `into_balance` pair, and the `coinWithBalance` import is gone.

File: packages/hashi/src/client.ts (hunk 9 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: After 82cd7467 this is moot — the coin/balance intermediates are gone, and the remaining `balance` result feeds a generated-binding call where a named variable reads fine. Resolving alongside the other two.

File: packages/hashi/src/client.ts (hunk 13 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 0942662b — `#batchGetObjects` is gone; verified all three transports chunk internally at 50 in their `getObjects`, so the local 500-chunking never changed what hit the wire anyway.

File: packages/hashi/src/client.ts (hunk 14 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 0942662b — the raw fetch is replaced with a `SuiGraphQLClient` query using real GraphQL variables ($sender/$type/$after), which also removes the string-interpolation injection surface. Happy to move this to the core event API once the gRPC/core event queries land.

File: packages/hashi/src/client.ts (hunk 6 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Fixed in 0942662b — the constructor now builds a `SuiGraphQLClient` (from `graphqlUrl` or the network default) instead of storing a bare URL, and the events query goes through it.

File: packages/hashi/src/client.ts (hunk 11 above)
Comment (Tzal3x) [reply packages/hashi/src/client.ts]: Yes, they were duplicates — fixed in a01838b6. Both estimators now call the real `tx.deposit`/`tx.requestWithdrawal` builders with dummy values, so the dry-run transaction can't drift from what users sign (the earlier `tx.balance` cleanup had to touch both copies, which proved the drift risk).

File: packages/hashi/src/client.ts (hunk 12 above)
Comment (hayes-mysten) [reply packages/hashi/src/client.ts:1399]: I really thought we has this, ignore me

**Comment by hayes-mysten** [comment]
@clud-bot rereview

**Review by hayes-mysten — APPROVED** [review]

File: packages/hashi/src/client.ts [hunk 20]
```diff
@@ -1459,37 +1443,15 @@ export class HashiClient {
...
-                    pageInfo { hasNextPage endCursor }
-                }
-            }`;
-
-			const res = await fetch(this.#graphqlUrl, {
-				method: 'POST',
-				headers: { 'Content-Type': 'application/json' },
-				body: JSON.stringify({ query }),
+			const result: GraphQLQueryResult<EventRequestIdsResult> = await this.#graphql.query({
+				query: EVENT_REQUEST_IDS_QUERY,
```
Comment (hayes-mysten) [inline packages/hashi/src/client.ts:1448]: Check out https://sdk.mystenlabs.com/sui/clients/graphql#custom-graphql-queries for the right way to use the graphql client, this avoids needing to manually type your result which is very error prone
